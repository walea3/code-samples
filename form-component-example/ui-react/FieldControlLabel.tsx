import { getDataRef } from '@nel-ui/js-helpers';
import cx from 'classnames';
import React, { ReactElement, ReactNode } from 'react';
import FormStyles from './FormStyles';

type OptionalFieldControlLabelProps =
  | {
      label: ReactNode;
      id: string;
      htmlFor: string;
    }
  | {
      label: never;
      id: never;
      htmlFor: never;
      fieldset: never;
    };

type AvailableAttr = React.AllHTMLAttributes<HTMLLabelElement | HTMLSpanElement>;
type BaseAttr = Pick<AvailableAttr, Exclude<keyof AvailableAttr, 'label'>> &
  OptionalFieldControlLabelProps;

export type FieldControlLabelProps = BaseAttr & {
  children: ReactElement;
  'data-ref'?: string;
  inline?: boolean;
};

export const FieldControlLabel = ({
  'data-ref': dataRef,
  label,
  htmlFor,
  className,
  id,
  inline = false,
  children,
  ...props
}: FieldControlLabelProps) => {
  if (!label) return children;

  return (
    <FormStyles>
      {({ classes }) => (
        <div
          data-ref={getDataRef(dataRef, 'fieldControlLabelContainer')}
          className={cx(className, classes.fieldControlLabelContainer, {
            [classes.inlineContainer]: inline,
          })}
          {...props}
        >
          <label
            data-ref={getDataRef(dataRef, 'fieldControlLabel')}
            className={classes.inputLabel}
            htmlFor={htmlFor}
            id={id}
          >
            {label}
          </label>

          {children}
        </div>
      )}
    </FormStyles>
  );
};
