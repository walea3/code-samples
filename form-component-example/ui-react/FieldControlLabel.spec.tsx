import { renderWithNEL } from '@nel-ui/tooling-react/test-utils';
import React from 'react';

import { FieldControlLabel, FieldControlLabelProps } from './FieldControlLabel';

describe('@renders', () => {
  const requiredProps = {
    children: <div>element</div>,
  };

  const label: FieldControlLabelProps['label'] = 'fieldControlLabel-label';

  const accessibilityProps: Partial<FieldControlLabelProps> = {
    label: 'fieldControlLabel-test-label',
    id: 'fieldControlLabel-test-id',
    htmlFor: 'fieldControlLabel-test-htmlFor',
  };

  const renderComponent = (props) =>
    renderWithNEL(<FieldControlLabel {...requiredProps} {...props} />);

  test.each`
    name                  | props
    ${'as default'}       | ${{}}
    ${'with label'}       | ${{ ...accessibilityProps, label }}
    ${'with id'}          | ${{ ...accessibilityProps, id: 'fieldControlLabel-id' }}
    ${'with htmlFor'}     | ${{ ...accessibilityProps, htmlFor: 'fieldControlLabel-htmlFor' }}
    ${'with className'}   | ${{ label, className: 'custom-className' }}
    ${'with inline'}      | ${{ label, inline: true }}
    ${'with data-ref'}    | ${{ label, 'data-ref': 'uid' }}
    ${'with other props'} | ${{ label, 'data-custom': 'custom-value' }}
  `('$name', ({ props }) => {
    expect(renderComponent(props)).toMatchSnapshot();
  });
});
