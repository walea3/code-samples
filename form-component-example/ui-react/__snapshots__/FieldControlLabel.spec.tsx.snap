// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@renders as default 1`] = `
<div>
  element
</div>
`;

exports[`@renders with className 1`] = `
<div
  class="custom-className nel-fieldControlLabelContainer"
  data-ref="fieldControlLabelContainer"
>
  <label
    class="nel-inputLabel"
    data-ref="fieldControlLabel"
  >
    fieldControlLabel-label
  </label>
  <div>
    element
  </div>
</div>
`;

exports[`@renders with data-ref 1`] = `
<div
  class="nel-fieldControlLabelContainer"
  data-ref="uid.fieldControlLabelContainer"
>
  <label
    class="nel-inputLabel"
    data-ref="uid.fieldControlLabel"
  >
    fieldControlLabel-label
  </label>
  <div>
    element
  </div>
</div>
`;

exports[`@renders with htmlFor 1`] = `
<div
  class="nel-fieldControlLabelContainer"
  data-ref="fieldControlLabelContainer"
>
  <label
    class="nel-inputLabel"
    data-ref="fieldControlLabel"
    for="fieldControlLabel-htmlFor"
    id="fieldControlLabel-test-id"
  >
    fieldControlLabel-test-label
  </label>
  <div>
    element
  </div>
</div>
`;

exports[`@renders with id 1`] = `
<div
  class="nel-fieldControlLabelContainer"
  data-ref="fieldControlLabelContainer"
>
  <label
    class="nel-inputLabel"
    data-ref="fieldControlLabel"
    for="fieldControlLabel-test-htmlFor"
    id="fieldControlLabel-id"
  >
    fieldControlLabel-test-label
  </label>
  <div>
    element
  </div>
</div>
`;

exports[`@renders with inline 1`] = `
<div
  class="nel-fieldControlLabelContainer nel-inlineContainer"
  data-ref="fieldControlLabelContainer"
>
  <label
    class="nel-inputLabel"
    data-ref="fieldControlLabel"
  >
    fieldControlLabel-label
  </label>
  <div>
    element
  </div>
</div>
`;

exports[`@renders with label 1`] = `
<div
  class="nel-fieldControlLabelContainer"
  data-ref="fieldControlLabelContainer"
>
  <label
    class="nel-inputLabel"
    data-ref="fieldControlLabel"
    for="fieldControlLabel-test-htmlFor"
    id="fieldControlLabel-test-id"
  >
    fieldControlLabel-label
  </label>
  <div>
    element
  </div>
</div>
`;

exports[`@renders with other props 1`] = `
<div
  class="nel-fieldControlLabelContainer"
  data-custom="custom-value"
  data-ref="fieldControlLabelContainer"
>
  <label
    class="nel-inputLabel"
    data-ref="fieldControlLabel"
  >
    fieldControlLabel-label
  </label>
  <div>
    element
  </div>
</div>
`;
